nuxeo-services/nuxeo-platform-types-core/src/main/java/org/nuxeo/ecm/platform/types/SubtypesJsonEnricher.java

```diff
@@ -0,0 +1,107 @@
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
+import org.apache.commons.lang3.BooleanUtils;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.io.marshallers.json.enrichers.AbstractJsonEnricher;
+import org.nuxeo.ecm.core.io.registry.reflect.Setup;
+import org.nuxeo.ecm.core.schema.SchemaManager;
+import org.nuxeo.ecm.platform.types.localconfiguration.UITypesConfigurationConstants;
+import org.nuxeo.runtime.api.Framework;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+
+/**
+ * Enrich {@link DocumentModel} JSON object with an array of the document types that can be created under the current
+ * document taking account the local configuration.
+ *
+ * @since 11.4
+ */
+@Setup(mode = SINGLETON, priority = REFERENCE)
+public class SubtypesJsonEnricher extends AbstractJsonEnricher<DocumentModel> {
+
+    public static final String NAME = "subtypes";
+
+    public SubtypesJsonEnricher() {
+        super(NAME);
+    }
+
+    @Override
+    public void write(JsonGenerator jg, DocumentModel enriched) throws IOException {
+        SchemaManager schemaManager = Framework.getService(SchemaManager.class);
+        Collection<String> defaultSubtypes = enriched.getDocumentType().getAllowedSubtypes();
+        Collection<String> subtypes = new HashSet<String>();
+
+        if (enriched.hasFacet(UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_FACET)) {
+            Boolean denyAllTypes = (Boolean) enriched.getPropertyValue(
+                    UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_DENY_ALL_TYPES_PROPERTY);
+            if (BooleanUtils.isNotTrue(denyAllTypes)) {
+                String[] allowedTypesProperty = (String[]) enriched.getPropertyValue(
+                        UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_ALLOWED_TYPES_PROPERTY);
+                String[] deniedTypesProperty = (String[]) enriched.getPropertyValue(
+                        UITypesConfigurationConstants.UI_TYPES_CONFIGURATION_DENIED_TYPES_PROPERTY);
+
+                List<String> allowedTypes = new ArrayList<>();
+                List<String> deniedTypes = new ArrayList<>();
+                if (allowedTypesProperty != null) {
+                    allowedTypes = Arrays.asList(allowedTypesProperty);
+                }
+                if (deniedTypesProperty != null) {
+                    deniedTypes = Arrays.asList(deniedTypesProperty);
+                }
+                if (!(allowedTypes.isEmpty() && deniedTypes.isEmpty())) {
+                    for (String subtype : defaultSubtypes) {
+                        if (!deniedTypes.contains(subtype)
+                                && (allowedTypes.isEmpty() || allowedTypes.contains(subtype))) {
+                            subtypes.add(subtype);
+                        }
+                    }
+                }
+            }
+        } else {
+            subtypes.addAll(defaultSubtypes);
+        }
```

Comment:
What about extracting this logic in a dedicated `protected` method? So that people can easily extend this enricher without having to copy/paste the whole code.